### PR TITLE
Independent JNI

### PR DESCRIPTION
**Pull request by mkitti**
This is a follow up to creating the JNI module. Here we the JNI initialization code into JNI itself as well as the associated globals. Further those globals are now converted to type associated `Ref`. Additionally the Julia interface is also simplified since `penv` no longer needs to be passed

### CODE REVIEW

**Comment by aviks** [comment]
Did appveyor take out Java from their images?

**Comment by mkitti** [comment]
Its my commit somehow. The master branch still passes when I rerun it.
https://ci.appveyor.com/project/aviks/javacall-jl-6c24s

**Comment by mkitti** [comment]
@aviks I fixed the appveyor issues.

Since I moved a lot of the low level functionality into the JNI submodule, I moved the Libdl import there as well. On Windows, mscrvt100.dll also needed to be loaded and that was causing the initial appveyor issues.

On further investigation into the appveyor configuration, I found that even when using the x86 platform, Julia was still trying to load the x64 JVM. By setting JAVA_HOME explicitly to the x86 JVM, and then fooling around with globals, I managed to get x86 init to work via 078278e. I don't know why that makes a difference.

**Comment by mkitti** [comment]
@aviks I intend to merge this this week.

File: src/JNI.jl [hunk 1]
```diff
@@ -19,6 +21,10 @@ include("jnienv.jl")
 const jniref = Ref(JNINativeInterface())
 global jnifunc
```
[the remark this answers is not in the document]
Comment (mkitti) [reply src/JNI.jl:22]: I kept `jnifunc` around mostly for backwards compatibility in case anyone else was accessing the global.

File: src/JNI.jl [hunk 2]
```diff
@@ -87,518 +93,596 @@ const JNI_EINVAL       = convert(Cint, -6)              #/* invalid arguments */
...
+    nOptions::Cint
+    options::Ptr{JavaVMOption}
+    ignoreUnrecognized::Cchar
+end
+
 function load_jni(penv::Ptr{JNIEnv})
-    jnienv = unsafe_load(penv)
+    global jnienv = unsafe_load(penv)
     jniref[] = unsafe_load(jnienv.JNINativeInterface_) #The JNI Function table
     global jnifunc = jniref[]
```
Comment (mkitti) [inline src/JNI.jl:115]: @aviks `jnifunc` is set here.

I also found that storing the `jnienv` as a global somehow makes x86 Windows work more smoothly. With that we have successful `JavaCall.init()` on x86 Windows. I wanted to see if I could understand why that makes a difference before I start cleaning up the globals.

File: src/JNI.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (aviks) [reply src/JNI.jl:22]: Ok cool. Maybe add a @deprecated_binding thing on it?

**Comment by mkitti** [comment]
@aviks The problem with `Base.@deprecated_binding` is that it declares a const at compile time for `jnifunc` which will be invalid. We could try to update it at runtime, but then this issues a warning about redefining a const.

I've left the deprecation warning, but trying to use `jnifunc` will fail now.

**Comment by mkitti** [comment]
To summarize the net changes:
1. Library loading and core JVM initialization, and destroy are moved into the JNI module. This means `__init__()` no longer loads the dynamic library. This is entirely done on `init()`
2. penv is no longer a global in JavaCall and passed by default to `ccall`. All JNI calls from JavaCall do not require a JNIEnv pointer to be passed since that is handled entirely within JNI module now. penv can still be passed in as a named argument.
3. JavaCall.findjvm() just returns a tuple of Strings describing the library or libraries to load rather than setting a global.
4. Tests no longer use `ccall` directly but now refer to the wrapped versions in the JNI module.

As a side effect and for unknown reasons, JavaCall.init() and some functions now work on x86 Windows. This is somehow related to storing `jnienv` as a global.

**Comment by mkitti** [comment]
Thanks for the review, @aviks . Merging now.

**Pull request merged**